src/client/dfuse/pil4dfs/int_dfs.c

```diff
@@ -788,51 +796,59 @@ parse_path(const char *szInput, int *is_target_path, dfs_obj_t **parent, char *i
 		}
 	}
 
+	/* Remove '/./'; Replace '//' with '/'; Remove '/' at the end of path. */
+	len = remove_dot_and_cleanup(full_path_parse, len);
+
+	/* standarlize and determine whether a path is a target path or not */
+
 	/* Assume full_path_parse[] = "/A/B/C/../D/E", it will be "/A/B/D/E" after
 	 * remove_dot_dot.
 	 */
 	rc = remove_dot_dot(full_path_parse, &len);
 	if (rc)
 		D_GOTO(out_err, rc = -rc);
 
-	/* Remove '/./'; Replace '//' with '/'; Remove '/' at the end of path. */
-	len = remove_dot_and_cleanup(full_path_parse, len);
-
+	/* determine whether the path contains any known dfs mount point */
 	idx_dfs = query_dfs_mount(full_path_parse);
 
 	if (idx_dfs >= 0) {
 		if (!daos_inited) {
```

Comment:
my point was to remove this daos_inited global variable and the global lock.
you can call init as many times as you want. daos internally will init once and will do refcounting on how many times you have inited. then you must call as many fini calls as you did init for daos to do the actual finalize.
so this is not expensive:
```
daos_init()
daos_init()
daos_init()
...
daos_fini()
daos_fini()
daos_fini()
```

but this is:
```
daos_init()
daos_fini()
daos_init()
daos_fini()
daos_init()
daos_fini()
```

however for the latter, you solution does not solve it here, and as i understand you are were attempting to solve the former? which should already be handled internally

Reply:
daos_fini() is only called once in the destructor function by the main thread. After daos_init() is called, it is not expensive to call daos_init() again.

https://github.com/daos-stack/daos/blob/master/src/client/api/init.c#L144-L149

accessing a global lock is still needed. With a global variable "daos_inited", we can avoid calling daos_init() multiple times.

Reply:
Thank you very much for your clarification! I added a counter "daos_init_cnt" to track the number of times daos_init() called. In the destruction phase, daos_fini() will be called daos_init_cnt times.